openwisp_firmware_upgrader/static/firmware-upgrader/js/upgrade-progress.js

```diff
@@ -407,3 +436,110 @@ function getFormattedDateTimeString(dateTimeString) {
   let dateTime = new Date(dateTimeString);
   return dateTime.toLocaleString();
 }
+
+function showCancelConfirmationModal(operationId) {
+  const $ = django.jQuery;
+
+  // Create modal if it doesn't exist
+  if ($("#ow-cancel-confirmation-modal").length === 0) {
+    createCancelConfirmationModal($);
+  }
+
+  // Set the operation ID and show the modal
+  $("#ow-cancel-confirmation-modal").data("operation-id", operationId);
+  $("#ow-cancel-confirmation-modal").removeClass("ow-hide");
+}
+
+function createCancelConfirmationModal($) {
+  const modalHtml = `
+    <div id="ow-cancel-confirmation-modal" class="ow-overlay ow-overlay-notification ow-overlay-inner ow-hide">
+      <div class="ow-dialog-notification ow-cancel-confirmation-dialog">
+        <span class="ow-dialog-close ow-dialog-close-x">&times;</span>
+        <div class="ow-cancel-confirmation-header">
+          <h2 class="ow-cancel-confirmation-title">${gettext("Stop upgrade operation")}</h2>
+        </div>
+        <div class="ow-cancel-confirmation-content">
+          <p>${gettext("Are you sure you want to cancel this upgrade operation?")}</p>
+        </div>
+        <div class="ow-dialog-buttons ow-cancel-confirmation-buttons">
+          <button class="ow-cancel-btn-confirm button default danger-btn">
+            ${gettext("Yes")}
+          </button>
+          <button class="ow-dialog-close button default">
+            ${gettext("No")}
+          </button>
+        </div>
+      </div>
+    </div>
+  `;
+
+  $("body").append(modalHtml);
+
+  // Close modal handlers
+  $("#ow-cancel-confirmation-modal .ow-dialog-close").on("click", function () {
+    $("#ow-cancel-confirmation-modal").addClass("ow-hide");
+  });
+
+  // Confirm cancellation handler
+  $("#ow-cancel-confirmation-modal .ow-cancel-btn-confirm").on("click", function () {
+    const operationId = $("#ow-cancel-confirmation-modal").data("operation-id");
+    $("#ow-cancel-confirmation-modal").addClass("ow-hide");
+    cancelUpgradeOperation(operationId);
+  });
+
+  // Close on escape key
+  $(document).on("keyup", function (e) {
+    if (e.keyCode === 27 && $("#ow-cancel-confirmation-modal").is(":visible")) {
+      $("#ow-cancel-confirmation-modal").addClass("ow-hide");
+    }
+  });
+
+  // Close on overlay click (outside dialog)
+  $("#ow-cancel-confirmation-modal").on("click", function (e) {
+    if (e.target === this) {
+      $(this).addClass("ow-hide");
+    }
+  });
+}
+
+function cancelUpgradeOperation(operationId) {
+  const $ = django.jQuery;
+
+  // Show loading overlay
+  $("#ow-loading").show();
+
+  $.ajax({
+    url: `/api/v1/firmware-upgrader/upgrade-operation/${operationId}/cancel/`,
+    type: "POST",
+    headers: {
+      "X-CSRFToken": $('input[name="csrfmiddlewaretoken"]').val(),
```

Comment:
I don't understand why this is needed? We don't generally include CSRF tokens in API calls. Is it really needed?

Reply:
https://github.com/openwisp/openwisp-firmware-upgrader/pull/329#discussion_r2320236689